### DISCUSSION TITLE
How to update image after file upload

### DISCUSSION THREAD

**Discussion by a-clare** [zauberzeug/nicegui#352] [Q&A] [answered]
I am trying to get an example going where the user uploads an image, I process the image using openCV, and then want to display the image (and eventually allow downloading image). 

However I cant seem to get the image to update correctly, and not sure where I need to initialize/create the ui.image().

I started with the https://nicegui.io/reference#file_upload example, and created a callback for `on_upload`

Should I put the `ui.image()` before ui.run(), and if yes how do I change the image after I process the image in my callback?

**Comment by rodja** [comment]
The `ui.image()` should be created at the position where you want to show it. In NiceGUI you define the structure of the layout by using `ui.row()`, `ui.column()`, `ui.card()`... and place the elements "inside" them by using contexts. So if you want to place the image beside the upload you would write

```py
with ui.rows():
    ui.upload(on_upload=...)
    ui.image()
```

To update the image you will need to hold it in a variable `result = ui.image()`. With that, you can simply change the url of the image by calling `result.set_source(url)`. The tricky part is the content of the `url` variable. You need to allow the browser to access the file with [`app.add_static_files(...)`](https://nicegui.io/reference#add_static_files_) or [custom FastAPI endpoints](https://github.com/zauberzeug/nicegui/discussions/28). Is this already working for you? Because you talk about "update" I guess you only see an outdated version of the image? That is caused by the browser cache. The browser only loads an image once for each url. To force a reload of the same image you need to change the url by adding some uniqueness for every update (for example by passing a query parameter). Say the image url is `/images/result.jpg`. Then you could write `result.set_source(f'{url}?time={time.time()}')` to append the current time, thereby making each url look new, but still providing the same filename. See 

https://github.com/zauberzeug/nicegui/blob/e79597e0b1bdc03218bb0005dfdfa33d435eae60/examples/opencv_webcam/main.py#L34

for a full example where this is used to present a video stream.

**Comment by falkoschindler** [comment, marked as answer]
Alternatively you write each uploaded image to a unique file to avoid caching:

```py
import os
import uuid
from nicegui import app, events, ui

os.makedirs('/tmp/images', exist_ok=True)
app.add_static_files('/images', '/tmp/images')

def handle_upload(event: events.UploadEventArguments) -> None:
    id = uuid.uuid4()
    with open(f'/tmp/images/{id}.png', 'wb') as f:
        f.write(event.content.read())
    image.set_source(f'/images/{id}.png')

ui.upload(on_upload=handle_upload)
image = ui.image()

ui.run()
```

**Reply by a-clare** [reply]
Thanks for the suggestion! This is working. 
But one follow up question, how can I reset the ui.image() after closing the web page?
I uploaded the image, displayed the image using your example, closed the web ui, re-opened the web ui and the previous image is still being displayed.

And a unrelated question, is it possible to let the users download the image (after I process it)

**Reply by rodja** [reply]
> how can I reset the ui.image() after closing the web page

Use [pages](https://nicegui.io/reference#page).

> it possible to let the users download the image

```py
ui.link('download', f'/images/{id}.png').props(f'download="{id}.png"')
```

**Reply by a-clare** [reply]
I appreciate all the help. 
I hope this is the last question:

I got the uploading image callback to trigger opening a new image display page.
However now I do not know how to access uuid generated in the image display function.

So I followed the example on modularization for pages and created a main, upload page, image page

```py
# main.py
from nicegui import ui
import upload_page

@ui.page('/')
def index_page() -> None:
  upload_page.content()

ui.run()
```

```py
# upload_page.py
import image_page
def content() -> None:
  ui.upload(on_upload=image_page.upload_and_display)
```

```py
# image_page.py
def upload_and_display(e: UploadEventArguments):
  id = uuid.uuid4()
  with open(f'/imgs/{id}.jpeg', 'wb') as f:
    f.write(e.content.read())
  
  # after finish writing the image go to the image display page
  # However I do not know how to access uuid now
  ui.open('/image')


@ui.page('/image')
def segmented_image():
    ui_image = ui.image()
    # here is where I do some image processing, I read the image /imgs/{id}.jpeg using opencv, do some processing, 
   # and write it out into /tmp/imgs/{id}.jpeg
    ui_image.set_source(f'/tmp/imgs/{id}.jpeg')
```

**Reply by a-clare** [reply]
I was able to get it by setting id as a global.
Not sure if there is a better way

**Reply by rodja** [reply]
You could pass the id as parameter to the image page: `ui.open(f'/image/{id}')`. Thanks to underlying FastAPI [the parameter can easily be received](https://nicegui.io/reference#pages_with_path_parameters):

```py
@ui.page('/image/{image_id}')
def segmented_image(image_id: str):
    ...
```
